DistroLauncher/upgrade_policy.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "stdafx.h"
+
+namespace internal
+{
+
+    bool starts_with(const std::wstring_view tested, const std::wstring_view start)
+    {
+        if (tested.size() < start.size()) {
+            return false;
+        }
+        auto mismatch = std::mismatch(start.cbegin(), start.cend(), tested.cbegin());
+        return mismatch.first == start.cend();
+    }
+
+    bool ends_with(const std::wstring_view tested, const std::wstring_view end)
+    {
+        if (tested.size() < end.size()) {
+            return false;
+        }
+        auto mismatch = std::mismatch(end.crbegin(), end.crend(), tested.crbegin());
+        return mismatch.first == end.crend();
+    }
+
+    std::wstring GetUpgradePolicy()
+    {
+        std::wstring_view name = DistributionInfo::Name;
+        if (name == L"Ubuntu") {
+            return L"lts";
+        }
+        if (starts_with(name, L"Ubuntu") && ends_with(name, L"LTS")) {
+            return L"never";
+        }
+        return L"normal";
+    }
+
+    void SetDefaultUpgradePolicyImpl()
+    {
+        namespace fs = std::filesystem;
+
+        const fs::path log{L"/var/log/upgrade-policy-changed.log"};
+        const fs::path policyfile{L"/etc/update-manager/release-upgrades"};
+
+        if (fs::exists(Oobe::WindowsPath(log))) {
+            return;
+        }
+
+        std::wstring regex = concat(L"s/Prompt=lts/Prompt=", GetUpgradePolicy(), L'/');
+        std::wstring sed = concat(L"sed -i ", std::quoted(regex), L' ', policyfile);
+        std::wstring date = concat(L"date --iso-8601=seconds > ", log);
+
+        std::wstring command = concat(L"bash -ec ", std::quoted(concat(sed, L" && ", date)));
+
+        DWORD errCode;
+        Sudo::WslLaunchInteractive(command.c_str(), FALSE, &errCode);
+    }
+
+}
+
+void SetDefaultUpgradePolicy()
+{
+    auto mutex = NamedMutex(L"upgrade-policy");
+    mutex.lock().and_then(internal::SetDefaultUpgradePolicyImpl);
```

Comment:
An alternative to my previous comment about filesystem exceptions might be setting a reasonable `or_else` function here. But that seems more complicated, unless you don't want to proceed in case of filesystem errors.

Reply:
Beware that `or_else` is called when we fail to switch user to root.

Throwing an exception during `and_then` will only revert the user to default and rethrow again; it will not call `or_else`.

I think the best is to have a fallback.